src/macros.rs

```diff
@@ -97,93 +95,88 @@ macro_rules! unsafe_impl {
+        fn is_bit_valid(candidate: MaybeValid<'_, Self>) -> bool {
             // SAFETY:
             // - The argument to `cast_unsized` is `|p| p as *mut _` as required
             //   by that method's safety precondition.
             // - The caller has promised that the cast results in an object of
             //   equal or lesser size.
-            // - The caller has promised that `$repr`'s alignment is less than
-            //   or equal to `Self`'s alignment.
             #[allow(clippy::as_conversions)]
             let candidate = unsafe { candidate.cast_unsized::<$repr, _>(|p| p as *mut _) };
-            // SAFETY:
-            // - The caller has promised that the referenced memory region will
-            //   contain a valid `$repr` for `'a`.
-            // - The memory may not be referenced by any mutable references.
-            //   This is a precondition of `is_bit_valid`.
-            // - The memory may not be mutated even via `UnsafeCell`s. This is a
-            //   precondition of `is_bit_valid`.
-            // - There must not exist any references to the same memory region
-            //   which contain `UnsafeCell`s at byte ranges which are not
-            //   identical to the byte ranges at which `T` contains
-            //   `UnsafeCell`s. This is a precondition of `is_bit_valid`.
-            let $candidate: &$repr = unsafe { candidate.as_ref() };
+
+            // SAFETY: The caller has promised that the referenced memory region
+            // will contain a valid `$repr`.
+            let candidate = unsafe { candidate.assume_valid() };
+
+            let $candidate = MaybeAligned::from_ptr(candidate);
             $is_bit_valid
         }
     };
-    (@method TryFromBytes ; |$candidate:ident: Ptr<$repr:ty>| $is_bit_valid:expr) => {
+    (@method TryFromBytes ; |$candidate:ident: MaybeValid<$repr:ty>| $is_bit_valid:expr) => {
         #[allow(clippy::missing_inline_in_public_items)]
         fn only_derive_is_allowed_to_implement_this_trait() {}
 
         #[inline]
-        unsafe fn is_bit_valid(candidate: Ptr<'_, Self>) -> bool {
+        fn is_bit_valid(candidate: MaybeValid<'_, Self>) -> bool {
             // SAFETY:
             // - The argument to `cast_unsized` is `|p| p as *mut _` as required
             //   by that method's safety precondition.
             // - The caller has promised that the cast results in an object of
             //   equal or lesser size.
-            // - The caller has promised that `$repr`'s alignment is less than
-            //   or equal to `Self`'s alignment.
             #[allow(clippy::as_conversions)]
             let $candidate = unsafe { candidate.cast_unsized::<$repr, _>(|p| p as *mut _) };
+
+            // SAFETY: The caller has promised that the referenced memory region
+            // will contain a valid `$repr`.
+            let $candidate = unsafe { $candidate.assume_as_initialized() };
```

Comment:
Isn't it the case that, when the caller provides a closure which takes a `MaybeValid`, they're *not* promising that any byte sequence is a valid `Self`?  IIUC, the only safety requirement here is that "the memory satisfies the initialization condition for `Self`" implies "the memory satisfies the initialization condition for `$repr`". I'm being vague about "initialization condition" because it's a pain to write down and we have another outstanding comment thread about the precise details.